### DISCUSSION TITLE
Transfer session between connections

### DISCUSSION THREAD

**Discussion by SunFulong** [wamp-proto/wamp-proto#487] [Q&A]
Hello,

I know Session's lifetime is will not early and later than WebSocket connection now.

But, if the WebSocket has max connection time limit, should we design to allow the Client save session before connection closed and restore it after established a new connection?

**Comment by ecorm** [comment]
This has already been brought up before but no consensus was achieved. See issue #260, as well as PR #264 which ended up being abandoned.

**Reply by SunFulong** [reply]
OK, I will implement it as an extension message, and wait for next round discuss.

**Comment by KSDaemon** [comment]
In general I think having such option in the protocol itself is a good option, in the meantime some implementations solve this problem in other ways:
* Wampy.js can automatically resubscribe/reregistrer upon reconnection (that no additional steps should be taken from the client side)
* Nexus allows setting cookies that can be used transparently during session establishment without the need to go full authorization cycle. 

Of course, this is not the same as session resumption it is a new session but this is enough for some cases.

**Reply by SunFulong** [reply]
As discussed earlier, I implemented it under AWS Server-less Application Model, their WebSocket Gateway cannot route pure array JSON messages, requires a keep-alive transaction of every 10 minutes, and ecah connection have 2 hours hard timeout.

But the current specification is not suitable, there is no heartbeat and recovery mechanism, and it does not support Chinese cipher algorithms, so I cannot use the 3rd party library and must implement it myself from scratch.

**Comment by oberstet** [comment]
fwiw, I think what has been discussed under "session resumption", and what some implementations currently provide only addresses part of the picture:

registrations/subscriptions "stay alive" rgd metadata (the routing information), but what about invocations/events which _did_ happen while the session was gone?

e.g. are events published to a topic while a session is offline then sent when the session is resumed? same for router-to-client side for subscriptions. similar for RPC ...

**Reply by SunFulong** [reply]
In my design, when client send a SAVE/PAUSE message to router, router will reply a GOODBYE message, then close the connection.

When client come back, will send a RESTORE/RESUME message, the session will continue.

During the SAVE/PAUSE time, all invocations and events are saved in a queue, when RESTORE/RESUME, send them to the client in sequence.

So, for each saved session, will set a timeout such as 30 seconds, after that, router should think the client is offline, and the refuse saved invocation messages as no such procedure.

**Comment by oberstet** [comment]
> their WebSocket Gateway cannot route pure array JSON messages,

IOW, they do not provide a WebSocket gateway, but an AWS proprietary API. Why do you want to enslave yourself? Just wondering ... I'm cool with that;) Freedom and choice is good. Including the freedom to enslave oneself ..

> it does not support Chinese cipher algorithms

As discussed extensively before, the spec could support it specifying those ciphers as options -- but of course you need implementations to follow.

And practically, who outside of China would care about those ciphers? Doesn't make sense. But if the Chinese government pays OSS projects to add the ciphers (under OSS license), that might work!

**Reply by SunFulong** [reply]
> they do not provide a WebSocket gateway, but an AWS proprietary API

I mean this link: https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-websocket-api.html

> Why do you want to enslave yourself?

Because it's the best option for my situation. I have about 80K active users, they are distributed in China Mainland, and the network exchange between different ISPs in China is very bad. Unfortunately, my city is at the end of the network, and a large number of users cannot connect to me fast and stably, so I have to choose a cloud computing provider to catch user traffic.

AWS will act the root/center router, each of my IDCs connect to AWS, register their procedure prefix, compute business locally and yield back to AWS.

> it does not support Chinese cipher algorithms

My mean is: current open source implementation not fit to me, so I need to implement it myself.

**Reply by oberstet** [reply]
> > they do not provide a WebSocket gateway, but an AWS proprietary API
> 
> I mean this link: https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-websocket-api.html

yes, and that's not a generic WebSocket gateway, but an AWS proprietory WebSocket API ..

**Reply by SunFulong** [reply]
> yes, and that's not a generic WebSocket gateway, but an AWS proprietory WebSocket API ..

OK, I got your mean, thank you!
